### PR TITLE
Expose landlock API

### PR DESCRIPTION
**Pull request by vlaci**
There is also a `landlock` package on PyPI implementing this, however I already implemented it by the time I found it

### CODE REVIEW

File: src/sandbox/mod.rs
Comment (qkaiser) [inline src/sandbox/mod.rs]: Out of curiosity, is `_py` underscored because it's unused in the function ?

Comment (vlaci) [reply src/sandbox/mod.rs]: Yes, but it also can be removed.

**Comment by qkaiser** [comment]
Made it working with the `landlock` branch in unblob (see https://github.com/onekey-sec/unblob/pull/597).

Had to use the v2 ABI otherwise we get into these kinds of issues due to the absence of `LANDLOCK_ACCESS_FS_REFER` :

 ```
 Traceback (most recent call last):
  File "/home/quentin/unblob/unblob/processing.py", line 639, in _extract_chunk
    if result := chunk.extract(inpath, extract_dir):
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/quentin/unblob/unblob/models.py", line 115, in extract
    return self.handler.extract(inpath, outdir)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/quentin/unblob/unblob/models.py", line 452, in extract
    return self.EXTRACTOR.extract(inpath, outdir)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/quentin/unblob/unblob/handlers/filesystem/romfs.py", line 312, in extract
    header.dump_fs()
  File "/home/quentin/unblob/unblob/handlers/filesystem/romfs.py", line 299, in dump_fs
    self.create_inode(inode)
  File "/home/quentin/unblob/unblob/handlers/filesystem/romfs.py", line 272, in create_inode
    self.create_hardlink(output_path, inode)
  File "/home/quentin/unblob/unblob/handlers/filesystem/romfs.py", line 263, in create_hardlink
    self.fs.create_hardlink(dst=output_path, src=target_path)
  File "/home/quentin/unblob/unblob/file_utils.py", line 594, in create_hardlink
    os.link(src, dst)
OSError: [Errno 18] Invalid cross-device link: '/tmp/pytest-of-quentin/pytest-2/test_all_handlers_filesystem_r0/fruits.romfs_extract/hlink_1_1' -> '/tmp/pytest-of-quentin/pytest-2/test_all_handlers_filesystem_r0/fruits.romfs_extract/dir_1/file_1.txt'
```

**Comment by qkaiser** [comment]
Two things to do before tagging as ready for review:

- [x] solve the `ModuleNotFoundError: No module named 'unblob_native.sandbox'` error by properly exposing it
- [x] add unit testing (unsupported platforms, different access checks)

**Comment by vlaci** [comment]
>     * [ ]  solve the `ModuleNotFoundError: No module named 'unblob_native.sandbox'` error by properly exposing it

I currently don't have any examples of it, but it worked for me by manually adding the submodule in `sys.modules`  
This issue contains that workaround suggestion together with suggesting just adding an `__init__.py`
https://github.com/PyO3/pyo3/issues/759

**Comment by qkaiser** [comment]
The module not found behavior is observed by pylance and pyright, but I confirmed that the module is actually in sys.module.

**Comment by qkaiser** [comment]
I have added a simple unit test that only runs on supported platforms (Linux x86-64). There is only one test because pytest runs all the test under the same process so calling `restrict_access` in one unit test would affect the behavior of all subsequent tests.

**Comment by vlaci** [comment]
> The module not found behavior is observed by pylance and pyright, but I confirmed that the module is actually in sys.module.

To solve that, we can add a `pyi` type stub: https://www.maturin.rs/project_layout.html#adding-python-type-information

> I have added a simple unit test that only runs on supported platforms (Linux x86-64). There is only one test because pytest runs all the test under the same process so calling restrict_access in one unit test would affect the behavior of all subsequent tests.

yeah, for separation a multiprocessing.Process or something alike is needed.

**Comment by qkaiser** [comment]
> To solve that, we can add a `pyi` type stub: https://www.maturin.rs/project_layout.html#adding-python-type-information

I leave that to you, this is alien to me.

> yeah, for separation a multiprocessing.Process or something alike is needed.

Do you think it's worth doing it ? I'd say the `landlock` Rust crate is probably well tested. Although we could add some tests on the Rust side of things.

File: tests/test_sandbox.py
Comment (qkaiser) [inline tests/test_sandbox.py]: We need to assert on this, but need to fix the function first me think.

File: src/sandbox/mod.rs
Comment (qkaiser) [inline src/sandbox/mod.rs:18]: We should improve on the signature of `restrict_access`. Ideally it should return a boolean value with 0 meaning it successfully restricted access, and 1 meaning the system is unsupported to be in line with the C API:

```
if (landlock_restrict_self(ruleset_fd, 0)) {
    perror("Failed to enforce ruleset");
    close(ruleset_fd);
    return 1;
}
close(ruleset_fd);
```

Comment (vlaci) [reply src/sandbox/mod.rs:18]: or raise an exception to be more pythonic. (which this implementation already does)

Comment (qkaiser) [reply src/sandbox/mod.rs:18]: It's not:

```
============================= test session starts ==============================
platform darwin -- Python 3.8.18, pytest-7.0.1, pluggy-1.0.0
rootdir: /Users/runner/work/unblob-native/unblob-native
collected 7 items

tests/test_math.py .....                                                 [ 71%]
tests/test_sandbox.py Fs                                                 [100%]

=================================== FAILURES ===================================
__________________________ test_unsupported_platform ___________________________

    @pytest.mark.skipif(platform.system() == "Linux", reason="Linux is supported.")
    def test_unsupported_platform():
        with pytest.raises(SandboxError):
>           restrict_access(AccessFS.read("/"))
E           Failed: DID NOT RAISE <class 'unblob_native.sandbox.SandboxError'>

tests/test_sandbox.py:14: Failed
------------------------------ Captured log call -------------------------------
WARNING  unblob_native.sandbox.sandbox_impl:unsupported.rs:6 Sandboxing FS access is unavailable on this system
=========================== short test summary info ============================
FAILED tests/test_sandbox.py::test_unsupported_platform - Failed: DID NOT RAI...
==================== 1 failed, 5 passed, 1 skipped in 0.08s ====================
```

Comment (qkaiser) [reply src/sandbox/mod.rs:18]: For unsupported platform I think a `SandboxError`should be raised with message `Sandboxing FS access is unavailable on this system` rather than logging the error.

Comment (vlaci) [reply src/sandbox/mod.rs:18]: yep, only the linux codepath can raise anything currently.

Comment (qkaiser) [reply src/sandbox/mod.rs:18]: changed the code so that the non-linux codepath would also raise an exception

**Comment by qkaiser** [comment]
@vlaci do you see anything blocking the merge of this PR ? Otherwise I'll approve it.

**Comment by vlaci** [comment]
> @vlaci do you see anything blocking the merge of this PR ? Otherwise I'll approve it.

Looks fine by me

**Review by qkaiser — APPROVED** [review]

**Pull request merged**
